### ISSUE TITLE
Cannot implicitly resolve function overload with generic values?

### ISSUE THREAD

**Issue by ComradeCow** [microsoft/TypeScript#1452]
I'm using tsc.exe version 1.3. 

Given the following code:

``` ts
    interface Observable<T> { 
    }

    interface StaticFuncs {
        unwrap<TValue>(value: Observable<TValue>): TValue;
        unwrap<TValue>(value: TValue): TValue;
    }

     var foo: StaticFuncs;

    var observable: Observable<number>;
    var value : number;
    var unwrappedObservable = foo.unwrap(observable);
    var unwrappedValue = foo.unwrap(value);

    function myFunction(param: number) {
    }

    myFunction(unwrappedObservable);
    myFunction(unwrappedValue); // the problem line
```

The TypeScript compiler returns two errors:

```
test.ts(20,1): error TS2082: Supplied parameters do not match any signature of call target. Could not apply type 'number' to argument 1 which is of type '{}'.
test.ts(20,1): error TS2087: Could not select overload for 'call' expression.
```

Curiously enough, if you switch the order of the 'unwrap' declarations, the compiler errors on line 19 and it says `Could not apply type 'number' to argument 1 which is of type 'Observable<number>'`. If I explicitly state `TValue` during the call `foo.unwrap<number>(observable)` the compiler is able to choose an overload correctly. The TypeScript tools (v1.1) in Visual Studio does not show an error for any of the above code.

I expected TypeScript to be able to infer the correct type for TValue for both overloads to `StaticFuncs.unwrap` regardless of the order they were declared in. I'm not sure if this is something supported by the spec or not, but this feels to me as something the language should support. This pattern is useful when developing [KnockoutJS](http://knockoutjs.com/documentation/introduction.html) applications in TypeScript. 

Let me know if there's any other information I can provide.
